### DISCUSSION TITLE
Events: Good idea to edit & save a custom field in afterSubmission event?

### DISCUSSION THREAD

**Discussion by mandrasch** [verbb/formie#1575] [Q&A] [answered]
Hi,

I'm currently learning more about the `afterSubmission` event (https://verbb.io/craft-plugins/formie/docs/developers/events#the-afterSubmission-event).

My use case is the following:
- Form is submitted
- I trigger an external API to create a new record
- I want to store the success state / error state of the API request
- I want to use a hidden field to store this data

My first idea:

```
Event::on(Submissions::class, Submissions::EVENT_AFTER_SUBMISSION, function(SubmissionEvent $event) {
   $submission = $event->submission;
   $submission->setFieldValue('apiCreateRecordSuccess', true); // or false
    Craft::$app->getElements()->saveElement($submission);
});
```

But I don't know if this is a good idea to use `saveElement()` in this event? 

Thanks very much in advance for any advice!
Regards, 
Matthias

**Comment by engram-design** [comment, marked as answer]
This might be a better task for a custom integration with Formie. But if you want to keep the quick-and-dirty path, that's okay!

That will be fine, no issue with that. You won't cause a recursion because `EVENT_AFTER_SUBMISSION` is only triggered via `formie/submissions/submit` action endpoint and not when saving the element at any given time. Otherwise, you'd end up in an infinite saving loop.

**Reply by mandrasch** [reply]
Thanks very much for the quick response! That was exactly what I was worried about 😬

**Comment by mandrasch** [comment]
HI @engram-design, just a quick follow up question about this:

Is it possible / a good idea to set the `isIncomplete` status back to 1 when I encounter errors with my external API service?

I inject my script like this now:

```
Event::on(Submissions::class, Submissions::EVENT_AFTER_SUBMISSION, function (SubmissionEvent $event) {

            // Beware: This is always successful on first form steps for multi page forms
            // (isIncomplete check is more important, see below)
            $success = $event->success;
            if (!$success) {
                return; // we only handle successful cases
            }

            /** @var Submission $submission */
            $submission = $event->submission;

            // Important - otherwise event will be triggered on first steps
            // https://verbb.io/craft-plugins/formie/docs/feature-tour/multi-page-forms
            if ($submission->isIncomplete) {
                return; // only handle completed submissions, not first steps
            }

            /** @var Form $form */
            $form = $submission->getForm();
            if ($form->handle !== 'mySpecialForm1') {
                return; // only handle the form with the correct handle, otherwise exit
            }
            
            // TODO: how to handle error case? show error, redirect user, ...?
            // throws exceptions in case of errors, IDEA would be:
            try{
                  $this->myService->sendFilesToExternalService($submission);
           }catch($e){
                $this->submission->setFieldValue('isIncomplete', 0);
               // TODO: save it
          }  
        });
```

Goal would be that user could have a second try submitting the form... (maybe API server had connection error)

(I'll check out how to develop a https://verbb.io/craft-plugins/formie/docs/developers/custom-integration in future, thanks very much for your advice along the way! Much appreciated! )

**Reply by engram-design** [reply]
That’s correct, setting `$submission->isIncomplete = true;` will prevent email notifications and integrations from firing. See [this line](https://github.com/verbb/formie/blob/aecaa3857d50db482749641a076b173bbdde1912/src/services/Submissions.php#L158)

You’ll also want to flag an error on the submission with `$submission->addError('general', 'some error')` which will prevent the submission from completing as per [this check](https://github.com/verbb/formie/blob/aecaa3857d50db482749641a076b173bbdde1912/src/controllers/SubmissionsController.php#L630C6-L630C6) after the `afterSubmission()` call
